### PR TITLE
consts and refactor, hash_address returns the hash

### PR DESCRIPTION
**Pull request by aldenml**

### CODE REVIEW

**Comment by aldenml** [comment]
Questions: do you think that this code, passing `address()`, can affect in a meaningful way the DHT?
```cpp
// TODO: 1 report the proper address of the router as the source IP of
// this vote of our external address, instead of the empty address
set_external_address(ip, source_router, address());
```

since the `cast_vote` will returns false and the last line in `session_impl::set_external_address` is
```
if (m_dht) m_dht->update_node_id();
```

**Comment by arvidn** [comment]
I would only imagine that would make a difference if you're behind a NAT and you don't have any other peers or trackers to talk to. The number of peers that tell you what your external IP would be much greater than a single router normally, so probably pretty marginal effect.

File: src/ip_voter.cpp [hunk 1]
```diff
@@ -111,8 +111,7 @@ namespace libtorrent
 
 		// this is the key to use for the bloom filters
 		// it represents the identity of the voter
-		sha1_hash k;
-		hash_address(source, k);
+		sha1_hash k = hash_address(source);
```
Comment (arvidn) [inline src/ip_voter.cpp]: now this could even be const :)

**Comment by aldenml** [comment]
Sure, I will put the consts

**Pull request merged**